### PR TITLE
Fix reading large files over SCP

### PR DESCRIPTION
**Pull request by Jakuje**
##### SUMMARY
The current SCP code does not handle large reads. Instead just allocates the large buffer fills up to 64k bytes and then writes the whole to the file (most likely the rest filled with junk).

The libssh returns the number of read bytes from `ssh_scp_read()`  and for practical reasons does not allocate the whole file in memory. The reads are capped to 64k B.

This PR fixes the read by both not allocating the entire file size in memory and reading it by chunks (that much max libssh chunk atm) and writing that to the local file.

##### ISSUE TYPE
- Bugfix Pull Request

##### ADDITIONAL INFORMATION

The attached test case is failing without the change, with errors like this:
```console
_______________________________________________________________________ test_get_large _______________________________________________________________________
[gw0] linux -- Python 3.12.3 /home/jjelen/devel/pylibssh/.tox/python/bin/python

dst_path = PosixPath('/tmp/pytest-of-jjelen/pytest-35/popen-gw0/test_get_large0/dst-file.txt')
src_path_large = PosixPath('/tmp/pytest-of-jjelen/pytest-35/popen-gw0/test_get_large0/large.txt'), ssh_scp = <pylibsshext.scp.SCP object at 0x7f3f83cffac0>
large_payload = b'&PE~~,!=1@F7RaUccg#)q-ru^~f8Zv/-{~tFE-\\>v2_b}i@`N/]_%*~{oD`"F`*4Ns]HK^lewILvC1O[]gFC9(27%MuX$s&24et<dRJ$"& zf_XIqm7...**<J_;?uI47vm\'/A"xi5a!kQXoL=M}ne;$E[B=EpasVbRO&\\zy7t,GK yu;l*$s<#P:sJLi@od#:&:m?8]9KFN4@QAF8hD, ~T@WNP;)%+>)TV\'ODi9'

    def test_get_large(dst_path, src_path_large, ssh_scp, large_payload):
        """Check that SCP file download works also for large files."""
>       ssh_scp.get(str(src_path_large), str(dst_path))

dst_path   = PosixPath('/tmp/pytest-of-jjelen/pytest-35/popen-gw0/test_get_large0/dst-file.txt')
large_payload = (b'&PE~~,!=1@F7RaUccg#)q-ru^~f8Zv/-{~tFE-\\>v2_b}i@`N/]_%*~{oD`"F`*4Ns]HK^le'
[...]
 b"?8]9KFN4@QAF8hD, ~T@WNP;)%+>)TV'ODi9")
src_path_large = PosixPath('/tmp/pytest-of-jjelen/pytest-35/popen-gw0/test_get_large0/large.txt')
ssh_scp    = <pylibsshext.scp.SCP object at 0x7f3f83cffac0>


tests/unit/scp_test.py:111: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise LibsshSCPException("Unexpected request: %s" % self._get_ssh_error_str())
E   pylibsshext.errors.LibsshSCPException: Unexpected request: b'ssh_scp_pull_request called under invalid state'


src/pylibsshext/scp.pyx:148: LibsshSCPException
```
(lines do not match)

Additionally, there is one more test case, that I suspected will be problematic, but it turned out it works just ok, so leaving for the coverage.

### CODE REVIEW

**Comment by webknjaz** [comment]
@Jakuje FYI the linting failures block this PR

File: docs/changelog-fragments/621.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fixed reading large files over SCP -- by :user:`Jakuje`
```
Comment (webknjaz) [inline docs/changelog-fragments/621.bugfix.rst]: Please, add a trailing period to the sentence. Also, it'd be nice to elaborate on this in a way that it would prompt the users of what changed in the behavior compared to the previous release. High-level, not need to be very detailed.

File: src/pylibsshext/scp.pyx [hunk 2]
```diff
@@ -122,6 +122,10 @@ cdef class SCP:
             size = libssh.ssh_scp_request_get_size(scp)
             mode = libssh.ssh_scp_request_get_permissions(scp)
 
+            # cap the buffer size to reasonable number -- libssh will not return the whole data at once anyway
+            read_buffer_size = size
+            if read_buffer_size > 65536:
+                read_buffer_size = 65536
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx]: There's an elegant helper for this:
```suggestion
            read_buffer_size = min(size, 65536)
```

File: src/pylibsshext/scp.pyx [hunk 3]
```diff
@@ -131,14 +135,17 @@ cdef class SCP:
...
-            # Read the file
-            rc = libssh.ssh_scp_read(scp, read_buffer, size)
-            if rc == libssh.SSH_ERROR:
-                raise LibsshSCPException("Error receiving file data: %s" % self._get_ssh_error_str())
-
-            py_file_bytes = read_buffer[:size]
             with open(local_file, "wb") as f:
-                f.write(py_file_bytes)
+                left = size
+                while left > 0:
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx]: Let's polish the vars to assign better meaning:
```suggestion
                remaining_bytes_to_read = size
                while remaining_bytes_to_read > 0:
```

File: src/pylibsshext/scp.pyx [hunk 4]
```diff
@@ -131,14 +135,17 @@ cdef class SCP:
...
+                    # Read the file
+                    read_bytes = libssh.ssh_scp_read(scp, read_buffer, left)
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx]:
```suggestion
                    read_bytes = libssh.ssh_scp_read(scp, read_buffer, remaining_bytes_to_read)
```

File: src/pylibsshext/scp.pyx [hunk 5]
```diff
@@ -131,14 +135,17 @@ cdef class SCP:
...
+                    if read_bytes == libssh.SSH_ERROR:
+                        raise LibsshSCPException("Error receiving file data: %s" % self._get_ssh_error_str())
+
+                    py_file_bytes = read_buffer[:read_bytes]
+                    f.write(py_file_bytes)
+                    left = left - read_bytes
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx]:
```suggestion
                    remaining_bytes_to_read -= read_bytes
```

File: tests/unit/scp_test.py [hunk 6]
```diff
@@ -57,3 +58,39 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+@pytest.fixture
+def src_path_large(tmp_path, large_payload):
+    """Return a remote path that is a file larger than 64k B, which is the most libssh can read at once."""
+    path = tmp_path / 'large.txt'
+    path.write_bytes(large_payload)
+    return path
+
+
+def test_get_large(dst_path, src_path_large, ssh_scp, large_payload):
+    """Check that SCP file download works also for large files."""
```
Comment (webknjaz) [inline tests/unit/scp_test.py]: I think this would be a more accurate high-level description:
```suggestion
    """Check that SCP file download gets over 64kB of data."""
```

**Comment by Jakuje** [comment]
The suggestions accepted, the linting should pass now.

File: docs/changelog-fragments/621.bugfix.rst [hunk 7]
```diff
@@ -0,0 +1 @@
+Downloading files larger thank 64kB over SCP no longer fails. -- by :user:`Jakuje`
```
Comment (webknjaz) [inline docs/changelog-fragments/621.bugfix.rst]: The trailing period should be after the byline, if it continues a sentence. Otherwise, a byline w/o a trailing period should go to a separate line with an empty-line separator.
```suggestion
Downloading files larger thank 64kB over SCP no longer fails -- by :user:`Jakuje`.
```

File: tests/unit/scp_test.py [hunk 8]
```diff
@@ -57,3 +58,45 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+
+def test_get_existing_local(dst_path_exists, src_path, ssh_scp, transmit_payload):
+    """Check that SCP file download works and overrides local file if it exists."""
+    ssh_scp.get(str(src_path), str(dst_path_exists))
+    assert dst_path_exists.read_bytes() == transmit_payload
+
+
+@pytest.fixture
+def large_payload():
+    """Generate a large test payload."""
```
Comment (webknjaz) [inline tests/unit/scp_test.py]: Let's be a bit more informative here:
```suggestion
    """Generate a large 65kB test payload."""
```

File: tests/unit/scp_test.py [hunk 9]
```diff
@@ -57,3 +58,45 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+    # buffer of random 1024 bytes -- just printable ones for easier debugging
+    rands = []
+    for _ in range(1024):
+        rnd = chr(random.randint(ord(' '), ord('~')))
+        rands.append(rnd)
+    rand = ''.join(rands)
+    # .. repeated 65 times
```
Comment (webknjaz) [inline tests/unit/scp_test.py]: This is already present in the variable name. Code comments shouldn't be used to repeat what the code does — just to explain motivation and unobvious considerations. For everything else — there's proper naming of vars and other identifiers.
```suggestion
```

File: tests/unit/scp_test.py [hunk 10]
```diff
@@ -57,3 +58,45 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
     """Check that SCP file download works."""
     ssh_scp.get(str(src_path), str(dst_path))
     assert dst_path.read_bytes() == transmit_payload
+
+
+@pytest.fixture
+def dst_path_exists(tmp_path):
+    """Return a local path that exists."""
+    path = tmp_path / 'non-existing.txt'
+    path.write_bytes('whatever'.encode())
```
Comment (webknjaz) [inline tests/unit/scp_test.py]: just use a byte string right away
```suggestion
    path.write_bytes(b'whatever')
```

File: tests/unit/scp_test.py [hunk 11]
```diff
@@ -57,3 +58,45 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+        rands.append(rnd)
+    rand = ''.join(rands)
+    # .. repeated 65 times
+    repeat = 65
+    return ''.join(rand for _ in range(repeat)).encode()
+
+
+@pytest.fixture
+def src_path_large(tmp_path, large_payload):
+    """Return a remote path that is a file larger than 64k B, which is the most libssh can read at once."""
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
    """Return a remote path that to a 65537 byte-sized file.
    
    Typical single-read chunk size is 64kB in ``libssh`` so
    the test needs a file that would overflow that to trigger
    the read loop.
    """
```

File: tests/unit/scp_test.py [hunk 12]
```diff
@@ -57,3 +58,45 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+    """Generate a large test payload."""
+    # buffer of random 1024 bytes -- just printable ones for easier debugging
+    rands = []
+    for _ in range(1024):
+        rnd = chr(random.randint(ord(' '), ord('~')))
+        rands.append(rnd)
+    rand = ''.join(rands)
+    # .. repeated 65 times
+    repeat = 65
+    return ''.join(rand for _ in range(repeat)).encode()
```
Comment (webknjaz) [inline tests/unit/scp_test.py]: I think, we could do better and use just one byte over the limit. It would make a better edge-case test.

Comment (webknjaz) [inline tests/unit/scp_test.py]: I realized there's a lot of conversions between integers, bytes and text strings back and forth here. So here's some refactoring to make it better. I'll post a series of comments with iterations of my refactoring and then, will commit it myself. Stay tuned!

This is the first improvement:

```suggestion
@pytest.fixture
def large_payload():
    """Generate a large 64kB+1B test payload."""
    # buffer of random 1024 bytes -- just printable ones for easier debugging
    rands = []
    for _ in range(1024):
        rnd = chr(random.randint(ord(' '), ord('~')))
        rands.append(rnd)
    rand = ''.join(rands)

    full_bytes_number = 64
    return ''.join(rand for _ in range(full_bytes_number)).encode() + rand[0]
```

Comment (webknjaz) [reply tests/unit/scp_test.py]:
```suggestion
@pytest.fixture
def large_payload():
    """Generate a large 64kB+1B test payload."""
    # buffer of random 1024 bytes -- just printable ones for easier debugging
    whitespace_position = ord(' ')
    tilda_position = ord('~')
    rands = []
    for _ in range(1024):
        rnd = chr(random.randint(whitespace_position, tilda_position))
        rands.append(rnd)
    rand = ''.join(rands)

    full_bytes_number = 64
    return ''.join(rand for _ in range(full_bytes_number)).encode() + rand[0]
```

Comment (webknjaz) [reply tests/unit/scp_test.py]:
```suggestion
from itertools import chain


@pytest.fixture
def large_payload():
    """Generate a large 65537 byte (64kB+1B) test payload."""
    # buffer of random 1024 bytes -- just printable ones for easier debugging
    whitespace_position = ord(' ')
    tilda_position = ord('~')
    rands = [random.randint(whitespace_position, tilda_position) for _ in range(1024)]

    full_bytes_number = 64
    return bytes(chain.from_iterable(rands for _ in range(full_bytes_number))) + rands[0].to_bytes(1)
```

Comment (webknjaz) [reply tests/unit/scp_test.py]:
```suggestion
@pytest.fixture
def large_payload():
    """Generate a large 65537 byte (64kB+1B) test payload."""
    # buffer of random 1024 bytes -- just printable ones for easier debugging
    whitespace_position = ord(' ')
    tilda_position = ord('~')
    rands = [random.randint(whitespace_position, tilda_position) for _ in range(1024)]

    full_bytes_number = 64
    return bytes(rands * full_bytes_number) + rands[0].to_bytes(1)
```

Comment (webknjaz) [reply tests/unit/scp_test.py]:
```suggestion
import string


@pytest.fixture
def large_payload():
    """Generate a large 65537 byte (64kB+1B) test payload."""
    random_char_kilobyte = [ord(random.choice(string.printable)) for _ in range(1024)]
    full_bytes_number = 64
    return bytes(random_char_kilobyte * full_bytes_number) + random_char_kilobyte[0].to_bytes(1)
```

File: tests/unit/scp_test.py [hunk 13]
```diff
@@ -57,3 +58,42 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+
+def test_get_existing_local(dst_path_exists, src_path, ssh_scp, transmit_payload):
+    """Check that SCP file download works and overrides local file if it exists."""
+    ssh_scp.get(str(src_path), str(dst_path_exists))
+    assert dst_path_exists.read_bytes() == transmit_payload
+
+
+import string
+
+
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
```

File: tests/unit/scp_test.py [hunk 14]
```diff
@@ -2,6 +2,7 @@
 
 """Tests suite for scp."""
 
+import random
```
Comment (webknjaz) [inline tests/unit/scp_test.py:6]:
```suggestion
import random
import string
```

File: tests/unit/scp_test.py [hunk 15]
```diff
@@ -57,3 +59,44 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
     """Check that SCP file download works."""
     ssh_scp.get(str(src_path), str(dst_path))
     assert dst_path.read_bytes() == transmit_payload
+
+
+@pytest.fixture
+def dst_path_exists(tmp_path):
```
Comment (webknjaz) [inline tests/unit/scp_test.py]: This name sounds like a check, but it's not. It should be renamed to reflect what it actually does.

Comment (webknjaz) [reply tests/unit/scp_test.py]: Moreover, it mentions that something exists, but the code uses `non-existing` in the name. Which one is it? I'm confused.

File: tests/unit/scp_test.py [hunk 16]
```diff
@@ -57,3 +59,44 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+    """Return a local path that exists."""
+    path = tmp_path / 'non-existing.txt'
+    path.write_bytes(b'whatever')
+    return path
+
+
+def test_get_existing_local(dst_path_exists, src_path, ssh_scp, transmit_payload):
+    """Check that SCP file download works and overrides local file if it exists."""
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
    """Check that SCP file download works and overwrites local file if it exists."""
```

File: tests/unit/scp_test.py [hunk 17]
```diff
@@ -57,3 +59,44 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
     """Check that SCP file download works."""
     ssh_scp.get(str(src_path), str(dst_path))
     assert dst_path.read_bytes() == transmit_payload
+
+
+@pytest.fixture
+def dst_path_exists(tmp_path):
+    """Return a local path that exists."""
+    path = tmp_path / 'non-existing.txt'
```
Comment (webknjaz) [inline tests/unit/scp_test.py]: oh...
```suggestion
    path = tmp_path / 'pre-existing-file.txt'
```

File: tests/unit/scp_test.py (hunk 15 above)
Comment (webknjaz) [reply tests/unit/scp_test.py]:
```suggestion
def pre_existing_file_path(tmp_path):
```

File: tests/unit/scp_test.py [hunk 18]
```diff
@@ -57,3 +59,44 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+    path.write_bytes(b'whatever')
+    return path
+
+
+def test_get_existing_local(dst_path_exists, src_path, ssh_scp, transmit_payload):
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
def test_get_existing_local(pre_existing_file_path, src_path, ssh_scp, transmit_payload):
```

File: tests/unit/scp_test.py [hunk 19]
```diff
@@ -57,3 +59,44 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+    """Check that SCP file download works and overrides local file if it exists."""
+    ssh_scp.get(str(src_path), str(dst_path_exists))
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
    ssh_scp.get(str(src_path), str(pre_existing_file_path))
```

File: tests/unit/scp_test.py [hunk 20]
```diff
@@ -57,3 +59,44 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+    assert dst_path_exists.read_bytes() == transmit_payload
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
    assert pre_existing_file_path.read_bytes() == transmit_payload
```

File: tests/unit/scp_test.py [hunk 21]
```diff
@@ -57,3 +59,44 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
     """Check that SCP file download works."""
     ssh_scp.get(str(src_path), str(dst_path))
     assert dst_path.read_bytes() == transmit_payload
+
+
+@pytest.fixture
+def pre_existing_file_path(tmp_path):
+    """Return a local path that exists."""
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
    """Return local path for a pre-populated file."""
```

File: src/pylibsshext/scp.pyx [hunk 22]
```diff
@@ -131,14 +133,17 @@ cdef class SCP:
...
-            rc = libssh.ssh_scp_read(scp, read_buffer, size)
-            if rc == libssh.SSH_ERROR:
-                raise LibsshSCPException("Error receiving file data: %s" % self._get_ssh_error_str())
-
-            py_file_bytes = read_buffer[:size]
             with open(local_file, "wb") as f:
-                f.write(py_file_bytes)
+                remaining_bytes_to_read = size
+                while remaining_bytes_to_read > 0:
+                    # Read the file
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx]: Comments repeating what the code does are pointless. Proper identifier names should be used instead.
```suggestion
```
Learning materials: https://www.youtube.com/watch?v=YklKUuDpX5c / https://www.youtube.com/watch?v=x-kB2o8sd5c / https://www.youtube.com/watch?v=1__lNTlj1_w.

File: src/pylibsshext/scp.pyx [hunk 23]
```diff
@@ -131,14 +133,17 @@ cdef class SCP:
...
+                    read_bytes = libssh.ssh_scp_read(scp, read_buffer, min(remaining_bytes_to_read, read_buffer_size))
+                    if read_bytes == libssh.SSH_ERROR:
+                        raise LibsshSCPException("Error receiving file data: %s" % self._get_ssh_error_str())
+
+                    py_file_bytes = read_buffer[:read_bytes]
+                    f.write(py_file_bytes)
+                    remaining_bytes_to_read -= read_bytes
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx:149]: I believe that this would be more correct:
```suggestion
                    remaining_bytes_to_read -= len(py_file_bytes)
```
The reason is that the slice operator may return fewer bytes, then you think in case the actual iterable is smaller:
```pycon
>>> a_list = [2, 3, 4]
>>> a_list
[2, 3, 4]
>>> a_list[:0]
[]
>>> a_list[:1]
[2]
>>> a_list[:2]
[2, 3]
>>> a_list[:3]
[2, 3, 4]
>>> a_list[:4]
[2, 3, 4]
>>> a_list[:5]
[2, 3, 4]
>>> a_list[:100500]
[2, 3, 4]
```

File: src/pylibsshext/scp.pyx [hunk 24]
```diff
@@ -131,14 +133,17 @@ cdef class SCP:
...
 
-            # Read the file
-            rc = libssh.ssh_scp_read(scp, read_buffer, size)
-            if rc == libssh.SSH_ERROR:
-                raise LibsshSCPException("Error receiving file data: %s" % self._get_ssh_error_str())
-
-            py_file_bytes = read_buffer[:size]
             with open(local_file, "wb") as f:
-                f.write(py_file_bytes)
+                remaining_bytes_to_read = size
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx]: While on it, let's move this var out to reduce the indentation:
```suggestion
            remaining_bytes_to_read = size
            with open(local_file, "wb") as f:
```

File: src/pylibsshext/scp.pyx [hunk 25]
```diff
@@ -122,7 +122,9 @@ cdef class SCP:
             size = libssh.ssh_scp_request_get_size(scp)
             mode = libssh.ssh_scp_request_get_permissions(scp)
 
-            read_buffer = <char *>PyMem_Malloc(size)
+            # cap the buffer size to reasonable number -- libssh will not return the whole data at once anyway
+            read_buffer_size = min(size, 65536)
```
Comment (webknjaz) [inline src/pylibsshext/scp.pyx]: Let's name this piece of data (though, I'd probably even move it into a module-level constant):
```suggestion
            default_read_buffer_size = 65536
            read_buffer_size = min(size, default_read_buffer_size)
```

Is there a constant within libssh that we could reuse instead of re-hardcoding it here?

Comment (webknjaz) [reply src/pylibsshext/scp.pyx]: Is it possible to rely on this https://github.com/canonical/libssh/blob/6f4ec9be/src/connector.c#L33?

File: tests/unit/scp_test.py [hunk 26]
```diff
@@ -57,3 +59,44 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+    ssh_scp.get(str(src_path), str(pre_existing_file_path))
+    assert pre_existing_file_path.read_bytes() == transmit_payload
+
+
+@pytest.fixture
+def large_payload():
+    """Generate a large 65537 byte (64kB+1B) test payload."""
+    random_char_kilobyte = [ord(random.choice(string.printable)) for _ in range(1024)]
+    full_bytes_number = 64
+    return bytes(random_char_kilobyte * full_bytes_number) + random_char_kilobyte[0].to_bytes(1)
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
    return bytes(random_char_kilobyte * full_bytes_number) + random_char_kilobyte[0].to_bytes(length=1, byteorder='big')
```

File: tests/unit/scp_test.py [hunk 27]
```diff
@@ -57,3 +59,44 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+    ssh_scp.get(str(src_path), str(pre_existing_file_path))
+    assert pre_existing_file_path.read_bytes() == transmit_payload
+
+
+@pytest.fixture
+def large_payload():
+    """Generate a large 65537 byte (64kB+1B) test payload."""
+    random_char_kilobyte = [ord(random.choice(string.printable)) for _ in range(1024)]
+    full_bytes_number = 64
+    return bytes(random_char_kilobyte * full_bytes_number) + random_char_kilobyte[0].to_bytes(length=1, byteorder='big')
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
    a_64kB_chunk = bytes(random_char_kilobyte * full_bytes_number)
    the_last_byte = random_choice(random_char_kilobyte).to_bytes(length=1, byteorder='big')
    return a_64kB_chunk + the_last_byte
```

File: tests/unit/scp_test.py [hunk 28]
```diff
@@ -57,3 +59,46 @@ def test_get(dst_path, src_path, ssh_scp, transmit_payload):
...
+    assert pre_existing_file_path.read_bytes() == transmit_payload
+
+
+@pytest.fixture
+def large_payload():
+    """Generate a large 65537 byte (64kB+1B) test payload."""
+    random_char_kilobyte = [ord(random.choice(string.printable)) for _ in range(1024)]
+    full_bytes_number = 64
+    a_64kB_chunk = bytes(random_char_kilobyte * full_bytes_number)
+    the_last_byte = random_choice(random_char_kilobyte).to_bytes(length=1, byteorder='big')
```
Comment (webknjaz) [inline tests/unit/scp_test.py]:
```suggestion
    the_last_byte = random.choice(random_char_kilobyte).to_bytes(length=1, byteorder='big')
```

**Comment by webknjaz** [comment]
@Jakuje hey, so I added a few improvements on top and posted a few more comments inline. I've now stopped pushing to the PR branch so feel free to clean it up/rebase/make more changes.

File: docs/changelog-fragments/621.bugfix.rst [hunk 29]
```diff
@@ -0,0 +1 @@
+Downloading files larger thank 64kB over SCP no longer fails -- by :user:`Jakuje`.
```
Comment (webknjaz) [inline docs/changelog-fragments/621.bugfix.rst]: just noticed a typo
```suggestion
Downloading files larger than 64kB over SCP no longer fails -- by :user:`Jakuje`.
```

File: src/pylibsshext/scp.pyx (hunk 25 above)
Comment (Jakuje) [reply src/pylibsshext/scp.pyx]: There is no constant for this. This one is for reading different chunks in different part of the protocol. The SCP has this hardcoded and given that we are deprecating SCP I do not expect this will ever change:

https://gitlab.com/libssh/libssh-mirror/-/blob/master/src/scp.c#L1025

Comment (Jakuje) [reply src/pylibsshext/scp.pyx]: I think we will have eventually the same problem with writing (but that is likely much less used in the context of ansible) so I will make a separate constant for this .

File: src/pylibsshext/scp.pyx (hunk 23 above)
Comment (Jakuje) [reply src/pylibsshext/scp.pyx:149]: Not sure if this is an issue as if we could not trust the consistency of the returned value `read_bytes` and the amount of read bytes, we are screwed anyway.

Given that we asked to read maximum of `read_buffer_size` and the buffer should have the same size we should not end up in the situation you describe. And if we do, we would create a different file that was on remote anyway so I would rather propose some defensive sanity check where we would fail instead of trying to save it with something like this. Committed a sanity check for the returned size that it is in the bounds we expect.

File: src/pylibsshext/scp.pyx (hunk 22 above)
Comment (Jakuje) [reply src/pylibsshext/scp.pyx]: Removed.

File: src/pylibsshext/scp.pyx (hunk 24 above)
Comment (Jakuje) [reply src/pylibsshext/scp.pyx]: Applied. Thanks!

File: src/pylibsshext/scp.pyx (hunk 23 above)
Comment (webknjaz) [reply src/pylibsshext/scp.pyx:149]: @Jakuje I see you added a check for getting *more* bytes which I suspect would be a protocol violation. But I was talking about the case when it returns *fewer* bytes, but we'd be removing more bytes than we actually got from the counter.

Imagine we have a file size of 5 and a buffer of 5 but the first invocation of `ssh_scp_read()` does not fill out the entire buffer and only receives 4 bytes. But we still set `read_buffer_size = read_buffer_size - buffer = 5 - 5 = 0` so after the next iteration never happens, and we never attempt getting that last byte.

**Comment by webknjaz** [comment]
This needs conflict resolution now.

File: src/pylibsshext/scp.pyx (hunk 23 above)
Comment (Jakuje) [reply src/pylibsshext/scp.pyx:149]: I got that, but I still do not see how this could happen with the current code nor how the proposed fix would fix that. We are subtracting the `read_bytes` from the `remaining_bytes_to_read` and that should always be equal to the `len(py_file_bytes)` unless the `read_bytes` would be larger than the `read_buffer_size`. Did you mistake the `read_buffer_size` and `read_bytes` variables?

**Comment by Jakuje** [comment]
Resolved the conflict and dropped the check for too large replies.

File: src/pylibsshext/scp.pyx (hunk 23 above)
Comment (webknjaz) [reply src/pylibsshext/scp.pyx:149]: Yeah, I got confused. I suppose `len(read_buffer[:read_bytes]) == read_bytes` is always true. Marking this thread as resolved.

**Review by webknjaz — APPROVED** [review]
Thanks!

**Comment by webknjaz** [comment]
Looks like one of the CI jobs got stuck with https://github.com/ansible/pylibssh/issues/557 so I restarted it: https://github.com/ansible/pylibssh/actions/runs/9699687288/job/26769948869?pr=621#step:15:152.

**Comment by webknjaz** [comment]
@Jakuje this has been released as a part of v1.2.2: https://github.com/ansible/pylibssh/releases/tag/v1.2.2 / https://github.com/ansible/pylibssh/discussions/626 / https://pypi.org/project/ansible-pylibssh/1.2.2/.

Enjoy! 🎉
